src/configuration.ts

```diff
@@ -1,180 +1,644 @@
 import * as appRoot from "app-root-path";
 import * as cluster from "cluster";
+import * as config from "config";
+import * as exp from "express";
 import * as fs from "fs-extra";
+import * as glob from "glob";
 import * as stringify from "json-stringify-safe";
+import * as _ from "lodash";
+import * as path from "path";
+import * as semver from "semver";
+
 import {
     HandleCommand,
     HandleEvent,
 } from "./index";
 import { Ingester, IngesterBuilder } from "./ingesters";
 import { logger } from "./internal/util/logger";
-import { obfuscateJson } from "./internal/util/string";
+import { guid, obfuscateJson } from "./internal/util/string";
 import { AutomationEventListener } from "./server/AutomationEventListener";
-import { AutomationServerOptions, RunOptions } from "./server/options";
 import { Maker } from "./util/constructionUtils";
 
-export interface Configuration extends AutomationServerOptions {
+/**
+ * Customize the express server configuration: For example to add custom routes
+ *
+ * Example:
+ *
+ * const newRouteCustomizer = (express: exp.Express, ...handlers: exp.RequestHandler[]) => {
+ *   express.get("/new-route", ...handlers, (req, res) => {
+ *       res.json({ key: "value" });
+ *   });
+ * }
+ */
+export type ExpressCustomizer = (express: exp.Express, ...handlers: exp.RequestHandler[]) => void;
 
-    commands?: Array<Maker<HandleCommand>>;
-    events?: Array<Maker<HandleEvent<any>>>;
-    ingesters?: Array<Ingester | IngesterBuilder>;
-    listeners?: AutomationEventListener[];
+/**
+ * Options for an automation node.
+ */
+export interface AutomationOptions {
+    /**
+     * Automation name.  If not given, the name is extracted from
+     * the package.json.
+     */
+    name?: string;
+    /**
+     * Automation version.  Must be a valid semantic version,
+     * https://semver.org/.  If not given, the version is extracted
+     * from the package.json.
+     */
+    version?: string;
+    /**
+     * Atomist teams this automation will be registered with.  Must be
+     * specified if groups is not specified.  Cannot be specified if
+     * groups is specified.
+     */
+    teamIds?: string[];
+    /**
+     * Groups this automation will be registered with.  Must be
+     * specified if teams is not specified.  Cannot be specified if
+     * teams is specified.
+     */
+    groups?: string[];
+    /**
+     * If events should be queued when the registration is not
+     * connected to the websocket, specificy "durable".  "ephemeral"
+     * is suited for testing and running locally.
+     */
+    policy?: "ephemeral" | "durable";
+    /**
+     * GitHub personal access token with repo and read:org scopes for
+     * all GitHub users and organizations associated with the Atomist
+     * teams this automation is registered in.
+     */
```

Comment:
aaaah this documentation is so beautiful

what if someone else associates their user? can no one else register an automation? (this is not a problem with this change, this stuff is just confusing)

Reply:
I see how this is confusing, particularly the user part.